dataplane/internal/engine/engine.go

```diff
@@ -233,45 +282,249 @@ func (e *Engine) AddLayer3PuntRule(ctx context.Context, portName string, ip []by
 	return nil
 }
 
-// AddIPRoute adds a route to the FIB with the input next hops.
-func (e *Engine) AddIPRoute(ctx context.Context, v4 bool, ip, mask []byte, vrf uint64, nextHops []*dpb.NextHop) error {
-	fib := fibV6Table
-	if v4 {
-		fib = fibV4Table
+// prefixToPrimitives returns the primitive types of the route prefix.
+// ip addr bytes, ip mask bytes, is ipv4, vrf id, error.
+func prefixToPrimitives(prefix *dpb.RoutePrefix) ([]byte, []byte, bool, uint64, error) {
+	var ip []byte
+	var mask []byte
+	var isIPv4 bool
+	vrf := prefix.GetVrfId()
+
+	switch pre := prefix.GetPrefix().(type) {
+	case *dpb.RoutePrefix_Str:
+		_, ipNet, err := net.ParseCIDR(pre.Str)
+		if err != nil {
+			return ip, mask, isIPv4, vrf, fmt.Errorf("failed to parse ip prefix: %v", err)
+		}
+		ip = ipNet.IP.To4()
+		mask = ipNet.Mask
+		isIPv4 = true
+		if ip == nil {
+			ip = ipNet.IP.To16()
+			mask = ipNet.Mask
+			isIPv4 = false
+		}
+	case *dpb.RoutePrefix_Mask:
+		ip = pre.Mask.Addr
+		mask = pre.Mask.Mask
+		switch len(ip) {
+		case net.IPv4len:
+			isIPv4 = true
+		case net.IPv6len:
+			isIPv4 = false
+		default:
+			return ip, mask, isIPv4, vrf, fmt.Errorf("invalid ip addr length")
+		}
+	default:
+		return ip, mask, isIPv4, vrf, fmt.Errorf("invalid prefix type")
+	}
+	return ip, mask, isIPv4, vrf, nil
+}
+
+// addNextHopList creates all the next hops from the message, then create a next hop group if there are multiple next hops.
```

Comment:
So technically each route never directly points to a nexthop ID, always NHG ID first. Is there a reason why you're providing direct access to NH IDs?

Reply:
incorrect, routes can point directly to next hops